### ISSUE TITLE
UserAgentApplication.acquireTokenSilient() always retrieves a new token.

### ISSUE THREAD

**Issue by Ra-Sedgwick** [AzureAD/microsoft-authentication-library-for-js#488]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Browser:
- [x ] Chrome version 70.0.3538.102 (Official Build) (64-bit)
- [ ] Firefox version XX
- [ ] IE version XX
- [ ] Edge version XX
- [ ] Safari version XX

## Library version

<pre><code>
Library version: 0.2.3

## Current behavior

UserAgentApplication.acquireTokenSilient() always retrieves a new token.

Im currently on v0.1.6. I updated to the current version 0.2.3 however the issues is persisting.

I have been unable to step into the msal code on the current version to debug, the debugger jumps into B2cAuthority.ts and then just returns.

However I have beenable to debug the in v 0.1.6.

I'm using the following method to obtain a token prior to api calls.

```javascript
public async getAuthToken(): Promise<string>  {
    
    let authToken = await this.userAgentApplication.acquireTokenSilent(this.appConfig.b2cScopes);
    
    this.logger.info(authToken);

    return authToken;    

  }
```

It looks like the problem is occuring in UserAgentApplication.getCachedToken() which is not returning the cached token failing on line 747.

I can see my cached token int the tokenCahcedItems array however the value of authenticationRequest.scopes appears be my cliendId so no match is found.

```javascript
  private getCachedToken(authenticationRequest: AuthenticationRequestParameters, user: User): CacheResult {
    let accessTokenCacheItem: AccessTokenCacheItem = null;
    const scopes = authenticationRequest.scopes;
    const tokenCacheItems = this._cacheStorage.getAllAccessTokens(this.clientId, user ? user.userIdentifier:null); //filter by clientId and user
    if (tokenCacheItems.length === 0) { // No match found after initial filtering
      return null;
    }

    const filteredItems: Array<AccessTokenCacheItem> = [];
    //if no authority passed
    if (!authenticationRequest.authority) {
      //filter by scope
      for (let i = 0; i < tokenCacheItems.length; i++) {
        const cacheItem = tokenCacheItems[i];
        const cachedScopes = cacheItem.key.scopes.split(" ");
        if (Utils.containsScope(cachedScopes, scopes)) {
          filteredItems.push(cacheItem);
        }
      }
```

## Expected behavior

I expect aquireTokenSilent() to return the same cached token if the token has not expired.

## Minimal reproduction of the problem with instructions

Log results of aquireTokenSilent().

</code></pre>

**Comment by nehaagrawal** [comment]
@Ra-Sedgwick  I am unable to reproduce this issue with the latest version(0.2.3). Can you please confirm what is the value of this.appConfig.b2cScopes. The value of authenticationRequest.scopes can be  cliendId only if you passed the clientID as the scope.

**Comment by Ra-Sedgwick** [comment]
@nehaagrawal 

Thank you for the assistance. 

The value of appConfig.b2cScopes is similar to the following 
```jon
"b2cScopes": [
    "https://example.com/webapi/read",
    "https://example.com/webapi/write"
  ],
```

I've also noticed that my app is restarting every time aquireTokenSilent() is called. I suspect this issue is related to #393

**Comment by ErikAllan** [comment]
Have you found any solution? I'm having the same problem :-(

**Comment by Ra-Sedgwick** [comment]
@ErikAllen I'm currently manuelly caching as a work around

 I strongly suspect this is tied to issue #393.   Once that issue is addressed I plan on looking back into this.

**Comment by navyasric** [comment]
There is a fix to return the cached token in the dev branch. Please give it a try and let us know if it resolves this issue. We will be releasing the fix soon.

**Comment by Kordonme** [comment]
I have this issue as well. The only thing I can tie this to at the moment, is when I haven't called `acquireTokenSilent()` for a long time. E.g. over the weekend. I'm using `localStorage` if that is useful.

Logging out and back in solves this issue. I'm not sure how to test if the token not being used for a period of time is the issue.

**Comment by MaDooSan** [comment]
@navyasric 
Same here. Not sure, but it looks like it might have to do with how the scopes for a single-tenant app are managed internally. 
The scopes are defined during MSAL Agent initialization as <client ID>/<scope> and the access token received after login is stored in the internal local cache. Calling acquireTokenSilent with the same scope syntax causes it to renew the token (and reload the page) because for some reason it cannot find the local token keyed by that scope name. When I make subsequent calls using acquireTokenSilent but with just the scope, i.e. <scope> instead of <client ID>/<scope>, it returns the local access token without renewal. However, this fails again when the token expires.
Looking forward to a fix.

**Comment by MaDooSan** [comment]
This may not exactly be a long-term fix but I modified the containsScope function so that it accepts both `clientID/scope` and `scope` formats:

`Utils.containsScope = function (cachedScopes, scopes) {
        cachedScopes = this.convertToLowerCase(cachedScopes);
        scopes = scopes.map((scope)=>{return scope.substring(scope.lastIndexOf("/") + 1)});
        return scopes.every(function (value) { return cachedScopes.indexOf(value.toString().toLowerCase()) >= 0; });
    }`

**Comment by xinyi-joffre** [comment]
+1 for @MaDooSan solution. For our app, we are going against Azure DevOps Rest API, which requires v1.0 app id in front of the scope when requesting the initial token: `499b84ac-1321-427f-aa17-267ca6975798/user_impersonation`. 

The response returned from Azure DevOps has only `user_impersonation` as the scope, and this appears in the cached key in local storage.

Later, when we go to acquireTokenSilent, we pass in the scope `499b84ac-1321-427f-aa17-267ca6975798/user_impersonation`, but the code checks if any of the cached scopes "contains" the request scope. Since `user_impersonation` does not contain the string `499b84ac-1321-427f-aa17-267ca6975798/user_impersonation`, it is not picking up the correct access token (so it fetches a new one each time). 

This scenario is still not yet fixed in the dev branch.

**Comment by raikoss** [comment]
I am experiencing the same issue. 

As I'm trying to retrieve a token before doing API calls, they take about 3-4 seconds (which is also pretty long) longer as MSAL fetches a new token even though it exists in the cache. Using the logger, it states that the token does not. I am using v0.2.4. It feels really counterintuitive to cache the tokens manually as a workaround, as MSAL should be doing this automatically.

**Comment by sameerag** [comment]
@raikoss msal does cache the tokens and looks up the cache for existing valid tokens for a request as long as the 'authority' and 'scopes' are matched in the original request (where the token is cached in the corresponding response).

We also recently fixed a bug that causes cache misses with authority.

We have recently gone through a redesign of the library's API surface. Please download our [latest preview package](https://www.npmjs.com/package/msal) or pull the [dev branch](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev) and try updating your code and see if the issue still persists. 

Please re-open this issue if it persists. We are now throwing error stack traces so we can understand better why your code is failing.

If you would like guidance on how to use the new version of the library, please review our wiki page [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/wiki/MSAL.js-1.0.0-preview-api-release).

**Comment by sameerag** [comment]
@Ra-Sedgwick Is this still seen on the later MSAL JS versions: 1.0.x?

**Comment by DarylThayil** [comment]
this should be fixed in msal 1.0 + 
1.1.1 is our most current version
